### ISSUE TITLE
Rolled back transaction has not been completed correctly" on Magento 2.1.15

### ISSUE THREAD

**Issue by adammprost** [magento/magento2#18752]
### Description
When a guest order is placed in Magento\Checkout\Model\GuestPaymentInformationManagement::savePaymentInformationAndPlaceOrder(), the following is executed: $salesConnection->beginTransaction(); $checkoutConnection->beginTransaction(), which creates a nested set of DB transactions (e.g. Magento\Framework\DB\Adapter\Pdo\Mysql::_transactionLevel exceeds 1).

Magento _is_ built to handle nested DB transactions, but the way GuestPaymentInformationManagement is architected, it's almost guaranteeing a critical exception. When $this->savePaymentInformation() is executed, the _transactionLevel is already at 2, which means that any rollbacks triggered inside savePaymentInformation() will lead to a Magento\Framework\DB\Adapter\AdapterInterface::ERROR_ROLLBACK_INCOMPLETE_MESSAGE exception message thrown. 

This is especially odd architecture when we compare it to synonymous class and method use for registered users: Magento\Checkout\Model\PaymentInformationManagement::savePaymentInformationAndPlaceOrder()

### Preconditions (*)
1. Magento 2.1.15
2. PHP 7.0

### Steps to reproduce (*)
1. Configure any online payment method with sandbox credentials.
2. Place a guest order with credit card details that trigger an error message from the payment processor.

### Expected result (*)
1. The customer is shown the "An error has occurred on the server.." message.
2. The exception log is empty.

### Actual result (*)
1. The customer is shown the "An error has occurred on the server.." message.
2. The exception log has a critical exception of "Rolled back transaction has not been completed correctly."

**Comment by magento-engcom-team** [comment]
Hi @adammprost. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@adammprost do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by kanduvisla** [comment]
I'm running into this issue as well after updating from Magento 2.1.15 to 2.2.6. Here's a stack trace that triggers the error in my production server (tested it first on acceptance, but there this issue doesn't happen :-/):

```
#0 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-catalog-inventory/Model/ResourceModel/Stock.php(174): Magento\Framework\DB\Adapter\Pdo\Mysql->beginTransaction()
#1 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-catalog-inventory/Model/StockManagement.php(151): Magento\CatalogInventory\Model\ResourceModel\Stock->correctItemsQty(Array, 0, '+')
#2 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-catalog-inventory/Observer/RevertQuoteInventoryObserver.php(67): Magento\CatalogInventory\Model\StockManagement->revertProductsSale(Array, 0)
#3 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Event/Invoker/InvokerDefault.php(72): Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver->execute(Object(Magento\Framework\Event\Observer))
#4 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Event/Invoker/InvokerDefault.php(60): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod(Object(Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver), Object(Magento\Framework\Event\Observer))
#5 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Event/Manager.php(66): Magento\Framework\Event\Invoker\InvokerDefault->dispatch(Array, Object(Magento\Framework\Event\Observer))
#6 /home/Vendor-webshop/releases/release-production-20181023-085210/generated/code/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager->dispatch('sales_model_ser...', Array)
#7 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-quote/Model/QuoteManagement.php(534): Magento\Framework\Event\Manager\Proxy->dispatch('sales_model_ser...', Array)
#8 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-quote/Model/QuoteManagement.php(406): Magento\Quote\Model\QuoteManagement->submitQuote(Object(Magento\Quote\Model\Quote\Interceptor), Array)
#9 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-quote/Model/QuoteManagement.php(364): Magento\Quote\Model\QuoteManagement->submit(Object(Magento\Quote\Model\Quote\Interceptor))
#10 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-quote/Model/GuestCart/GuestCartManagement.php(87): Magento\Quote\Model\QuoteManagement->placeOrder('17289', NULL)
#11 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-checkout/Model/GuestPaymentInformationManagement.php(106): Magento\Quote\Model\GuestCart\GuestCartManagement->placeOrder('5y5G97NRZinMKGJ...')
#12 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Checkout\Model\GuestPaymentInformationManagement->savePaymentInformationAndPlaceOrder('5y5G97NRZinMKGJ...', 'giel.berkers@is...', Object(Magento\Quote\Model\Quote\Payment), Object(Magento\Quote\Model\Quote\Address))
#13 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Interception/Interceptor.php(138): Vendor\BpostShipping\Rewrite\Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor->___callParent('savePaymentInfo...', Array)
#14 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Interception/Interceptor.php(153): Vendor\BpostShipping\Rewrite\Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor->Magento\Framework\Interception\{closure}('5y5G97NRZinMKGJ...', 'giel.berkers@is...', Object(Magento\Quote\Model\Quote\Payment), Object(Magento\Quote\Model\Quote\Address))
#15 /home/Vendor-webshop/releases/release-production-20181023-085210/generated/code/Vendor/BpostShipping/Rewrite/Magento/Checkout/Model/GuestPaymentInformationManagement/Interceptor.php(39): Vendor\BpostShipping\Rewrite\Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor->___callPlugins('savePaymentInfo...', Array, Array)
#16 [internal function]: Vendor\BpostShipping\Rewrite\Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor->savePaymentInformationAndPlaceOrder('5y5G97NRZinMKGJ...', 'giel.berkers@is...', Object(Magento\Quote\Model\Quote\Payment), Object(Magento\Quote\Model\Quote\Address))
#17 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(95): call_user_func_array(Array, Array)
#18 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/module-webapi/Controller/Rest.php(244): Magento\Webapi\Controller\Rest\SynchronousRequestProcessor->process(Object(Magento\Framework\Webapi\Rest\Request\Proxy))
#19 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#20 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor->___callParent('dispatch', Array)
#21 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#22 /home/Vendor-webshop/releases/release-production-20181023-085210/generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(26): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins('dispatch', Array, Array)
#23 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#24 /home/Vendor-webshop/releases/release-production-20181023-085210/vendor/magento/framework/App/Bootstrap.php(257): Magento\Framework\App\Http->launch()
#25 /home/Vendor-webshop/releases/release-production-20181023-085210/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#26 {main}{"message":"Er is een fout opgetreden op de server. Probeer de bestelling opnieuw te plaatsen."}
```

**Comment by kanduvisla** [comment]
@adammprost : did you found a (temporary) solution to this problem?

**Comment by kanduvisla** [comment]
I found the issue in my case: I set the customer date of birth on the quote object, but the format was incorrect.

If anyone else bumps into this issue: your problem might be something else. This error (with a very generic "something went wrong" message and not a very clear stack trace) is caused if a transaction has been rolled back and is being re-opened later on.

The easiest way to debug this is by checking what process is triggering you `rollback()`-method. So if you add a very simple command at the beginning of `\Magento\Framework\DB\Adapter\Pdo\Mysql::rollBack()` you can debug this in your specific case:

    echo (new \Exception())->getTraceAsString();

However, I do agree with @adammprost his comment that this setup can lead to some nasty situations that are difficult to debug.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96154`, `MAGETWO-96155` were created

**Comment by atsareva** [comment]
Hi! Any progress for this issue? Client said that he got a lot of messages from customers regarding this issue. I think it's urgent because when customers get 'An error has occurred on the server..', they don't know what they should do.

**Comment by jkuczek15** [comment]
Has anyone found a temporary solution to this? Still having problems when placing an order.

**Comment by weverson83** [comment]
I find out how to reproduce it:
It happens when a roll back is called and you catch the exception.

```
class RollbackExceptionTest extends \PHPUnit\Framework\TestCase
{
    /**
     * @magentoDbIsolation enabled
     */
    public function testRollbackException()
    {
        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();

        $resource = $objectManager->get(\Magento\Eav\Model\ResourceModel\Entity\Attribute\Set::class);
        $attributeSet = $this->getAttrSetInstance($objectManager);

        //Save new entity
        $resource->save($attributeSet);

        //Saving the second time throws AlreadyExistsException
        try {
            $attributeSet = $this->getAttrSetInstance($objectManager);
            $resource->save($attributeSet);
        } catch (\Magento\Framework\Exception\AlreadyExistsException $exception) {
            $this->assertTrue(true, 'Exception is thrown');
        }
      
        //Saving the 3rd time, throws Rolled back transaction has not been completed correctly
        try {
            $attributeSet = $this->getAttrSetInstance($objectManager);
            $resource->save($attributeSet);
        } catch (\Magento\Framework\Exception\AlreadyExistsException $exception) {
            $this->assertTrue(true, 'Exception is thrown');
        }
    }

    private function getAttrSetInstance(\Magento\Framework\App\ObjectManager $objectManager)
    {
        $factory = $objectManager->get(\Magento\Eav\Model\Entity\Attribute\SetFactory::class);
        $attributeSet = $factory->create();

        $attributeSet->setEntityTypeId(4);
        $attributeSet->setName('attrSetTest');

        return $attributeSet;
    }
}
```

To solve it, check in your code where those kind of exceptions are happening and fix it.

**Comment by josenacasabogiii** [comment]
I also had this issue after migrating from Magento 1.9.3 to Magento 2.3. Not sure how to fix it. Is there an update or quick fix to this one? Thank you!!

**Comment by josenacasabogiii** [comment]
I don't understand this error 

An exception occurred on 'sales_model_service_quote_submit_failure' event: Rolled back transaction has not been completed correctly.
#0 /home/porcel12/public_html/products-new/vendor/magento/framework/Model/ResourceModel/AbstractResource.php(65): Magento\Framework\DB\Adapter\Pdo\Mysql-&gt;beginTransaction()
#1 /home/porcel12/public_html/products-new/vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php(397): Magento\Framework\Model\ResourceModel\AbstractResource-&gt;beginTransaction()
#2 /home/porcel12/public_html/products-new/vendor/magento/framework/Model/AbstractModel.php(648): Magento\Framework\Model\ResourceModel\Db\AbstractDb-&gt;save(Object(Magento\Indexer\Model\Indexer\State))
#3 /home/porcel12/public_html/products-new/vendor/magento/module-indexer/Model/Indexer.php(447): Magento\Framework\Model\AbstractModel-&gt;save()
#4 /home/porcel12/public_html/products-new/vendor/magento/module-indexer/Model/Indexer/DependencyDecorator.php(271): Magento\Indexer\Model\Indexer-&gt;reindexList(Array)
#5 /home/porcel12/public_html/products-new/vendor/magento/framework/Indexer/AbstractProcessor.php(69): Magento\Indexer\Model\Indexer\DependencyDecorator-&gt;reindexList(Array)
#6 /home/porcel12/public_html/products-new/vendor/magento/module-inventory-sales/Observer/CatalogInventory/RevertQuoteInventoryObserver.php(49): Magento\Framework\Indexer\AbstractProcessor-&gt;reindexList(Array)
#7 /home/porcel12/public_html/products-new/vendor/magento/framework/Event/Invoker/InvokerDefault.php(72): Magento\InventorySales\Observer\CatalogInventory\RevertQuoteInventoryObserver-&gt;execute(Object(Magento\Framework\Event\Observer))
#8 /home/porcel12/public_html/products-new/vendor/magento/framework/Event/Invoker/InvokerDefault.php(60): Magento\Framework\Event\Invoker\InvokerDefault-&gt;_callObserverMethod(Object(Magento\InventorySales\Observer\CatalogInventory\RevertQuoteInventoryObserver), Object(Magento\Framework\Event\Observer))
#9 /home/porcel12/public_html/products-new/vendor/magento/framework/Event/Manager.php(66): Magento\Framework\Event\Invoker\InvokerDefault-&gt;dispatch(Array, Object(Magento\Framework\Event\Observer))
#10 /home/porcel12/public_html/products-new/generated/code/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager-&gt;dispatch('sales_model_ser...', Array)
#11 /home/porcel12/public_html/products-new/vendor/magento/module-quote/Model/QuoteManagement.php(536): Magento\Framework\Event\Manager\Proxy-&gt;dispatch('sales_model_ser...', Array)
#12 /home/porcel12/public_html/products-new/vendor/magento/module-quote/Model/QuoteManagement.php(406): Magento\Quote\Model\QuoteManagement-&gt;submitQuote(Object(Magento\Quote\Model\Quote\Interceptor), Array)
#13 /home/porcel12/public_html/products-new/vendor/magento/module-quote/Model/QuoteManagement.php(364): Magento\Quote\Model\QuoteManagement-&gt;submit(Object(Magento\Quote\Model\Quote\Interceptor))
#14 /home/porcel12/public_html/products-new/vendor/magento/module-quote/Model/GuestCart/GuestCartManagement.php(87): Magento\Quote\Model\QuoteManagement-&gt;placeOrder('51', NULL)
#15 /home/porcel12/public_html/products-new/vendor/magento/module-checkout/Model/GuestPaymentInformationManagement.php(107): Magento\Quote\Model\GuestCart\GuestCartManagement-&gt;placeOrder('msTSRa3wPxuSgN8...')
#16 /home/porcel12/public_html/products-new/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Checkout\Model\GuestPaymentInformationManagement-&gt;savePaymentInformationAndPlaceOrder('msTSRa3wPxuSgN8...', 'jnacasabog3@gma...', Object(Magento\Quote\Model\Quote\Payment\Interceptor), Object(Magento\Quote\Model\Quote\Address))
#17 /home/porcel12/public_html/products-new/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor-&gt;___callParent('savePaymentInfo...', Array)
#18 /home/porcel12/public_html/products-new/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor-&gt;Magento\Framework\Interception\{closure}('msTSRa3wPxuSgN8...', 'jnacasabog3@gma...', Object(Magento\Quote\Model\Quote\Payment\Interceptor), Object(Magento\Quote\Model\Quote\Address))
#19 /home/porcel12/public_html/products-new/generated/code/Magento/Checkout/Model/GuestPaymentInformationManagement/Interceptor.php(26): Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor-&gt;___callPlugins('savePaymentInfo...', Array, Array)
#20 [internal function]: Magento\Checkout\Model\GuestPaymentInformationManagement\Interceptor-&gt;savePaymentInformationAndPlaceOrder('msTSRa3wPxuSgN8...', 'jnacasabog3@gma...', Object(Magento\Quote\Model\Quote\Payment\Interceptor), Object(Magento\Quote\Model\Quote\Address))
#21 /home/porcel12/public_html/products-new/vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(95): call_user_func_array(Array, Array)
#22 /home/porcel12/public_html/products-new/vendor/magento/module-webapi/Controller/Rest.php(188): Magento\Webapi\Controller\Rest\SynchronousRequestProcessor-&gt;process(Object(Magento\Framework\Webapi\Rest\Request\Proxy))
#23 /home/porcel12/public_html/products-new/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest-&gt;dispatch(Object(Magento\Framework\App\Request\Http))
#24 /home/porcel12/public_html/products-new/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor-&gt;___callParent('dispatch', Array)
#25 /home/porcel12/public_html/products-new/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor-&gt;Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#26 /home/porcel12/public_html/products-new/generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(26): Magento\Webapi\Controller\Rest\Interceptor-&gt;___callPlugins('dispatch', Array, Array)
#27 /home/porcel12/public_html/products-new/vendor/magento/framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor-&gt;dispatch(Object(Magento\Framework\App\Request\Http))
#28 /home/porcel12/public_html/products-new/vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http-&gt;launch()
#29 /home/porcel12/public_html/products-new/index.php(46): Magento\Framework\App\Bootstrap-&gt;run(Object(Magento\Framework\App\Http\Interceptor))
#30 {main}

**Comment by david-fuehr** [comment]
Hi @josenacasabogiii,

you used the first commit on this issue. I just added a second commit and created the pull request - try to use these changes.

The error message a stack trace you provide just shows the consecutive exception, that happens in `RevertQuoteInventoryObserver.php`.

The `exception.log` should already contain the preceding exception message. Please provide this, if you need further assistance on identifying your problem.

**Comment by dazz397** [comment]
I upgraded from 2.1.10 to 2.2.7 and this started happening. It would only happen when there was a payment decline. I am using the authorize.net payment gateway from paradox labs.  I am not sure it's related but the exception being caught that causes this error is a message from that module. You can replicate it by making a payment using CC: 4444333322221111 CVV:111 EXP:0120.

Now for the fix.

In /vendor/magento/module-checkout/Model/GuestPaymentInformationManagement.php

Go to function savePaymentInformationAndPlaceOrder and comment out the following lines:

//$salesConnection = $this->connectionPull->getConnection('sales');
//$checkoutConnection = $this->connectionPull->getConnection('checkout');
//$salesConnection->beginTransaction();
//$checkoutConnection->beginTransaction();

//$salesConnection->commit();
//$checkoutConnection->commit();

//$salesConnection->rollBack();
//$checkoutConnection->rollBack();

I did a diff with the same file from 2.1.10 and it looks like the database transaction code did not exist in that file so it is something new that was added causing the problem. For now, I just commented it out and everything works.

**Comment by magento-engcom-team** [comment]
Hi @adammprost. Thank you for your report. 
The issue has been fixed in magento/magento2#21697 by @david-fuehr in 2.3-develop branch
Related commit(s):
 - [9480e2a8efbf2e013c0372af211b8e4a6cffa6c2](https://github.com/magento/magento2/commit/9480e2a8efbf2e013c0372af211b8e4a6cffa6c2)
 - [04f60d5d27f661b2a33c1f6db2ef511ab5ff3c2e](https://github.com/magento/magento2/commit/04f60d5d27f661b2a33c1f6db2ef511ab5ff3c2e)
 - [6b0ebf49055d0ee4bb371ffc23186eef2eb02fb7](https://github.com/magento/magento2/commit/6b0ebf49055d0ee4bb371ffc23186eef2eb02fb7)
 - [59771c40c1a2aa857d3b64b0905aa67c1705a1c9](https://github.com/magento/magento2/commit/59771c40c1a2aa857d3b64b0905aa67c1705a1c9)

The fix will be available with the upcoming 2.3.2 release.

**Comment by leoquijano** [comment]
Hello @magento-engcom-team, I see that there's a fix for the 2.3.x branch. Is there a fix to be released in the 2.2.x line?

**Comment by david-fuehr** [comment]
Hi @leoquijano,
@larsroettig just created the backport for 2.1: https://github.com/magento/magento2/pull/22037

**Comment by leoquijano** [comment]
Thanks @david-fuehr. I see there's a 2.2 version as well, but it's not an official one yet, right?

It's probably a good idea to have the final release number announced here.

**Comment by rhoerr** [comment]
Great to see movement here, but that's not really fixing the issue, is it?

It doesn't fix the possible rollback errors since 2.1.13/2.2.4, it just improves the exception logging around it when it does occur. It's an improvement, but not actually a solution.

**Comment by david-fuehr** [comment]
Hi @rhoerr,

as far as I know, there is no one root exception that causes this incomplete transaction rollback. So I would not know how to fix this generally.

The consecutive exception is thrown when a transaction rollback is triggered during placement of the order (e.g. `\Magento\Sales\Model\OrderRepository::save`):

```
/* pseudo stack trace */
\Magento\Framework\DB\Adapter\Pdo\Mysql::beginTransaction
\Magento\CatalogInventory\Model\ResourceModel\Stock::correctItemsQty
\Magento\CatalogInventory\Model\StockManagement::revertProductsSale
\Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver::execute
```

If I find the time before someone else fixes this, I will address this issue in another PR.

**Comment by magento-engcom-team** [comment]
Hi @adammprost. Thank you for your report. 
The issue has been fixed in magento/magento2#22037 by @larsroettig in 2.2-develop branch
Related commit(s):
 - [d91acbe552375ce794e97894644afe7851ca201f](https://github.com/magento/magento2/commit/d91acbe552375ce794e97894644afe7851ca201f)
 - [91410778d09b45bd2582683b88ff6169ac7898eb](https://github.com/magento/magento2/commit/91410778d09b45bd2582683b88ff6169ac7898eb)
 - [0168b1fdda9ab5ae770fb69ee519948a0529fd18](https://github.com/magento/magento2/commit/0168b1fdda9ab5ae770fb69ee519948a0529fd18)

The fix will be available with the upcoming 2.2.9 release.

**Comment by rhoerr** [comment]
Belated response: Reopening this. The above changes do not resolve the issue. They improve error handling around the critical rollback exception, but do not fix it being thrown in the first place, which never happened on checkout prior to 2.2.4.

I've created a minimum proof of concept to reproduce the error (tested on 2.3.2): https://github.com/rhoerr/proofofconcept-github18752

---

The issue is caused by the following changes to GuestPaymentInformationManagement in 2.2.4: https://github.com/magento/magento2/commit/e76430179fd9679f448883d05be40ae5423dd067#diff-5a8b5ff2d116c1abadb36cf6940a8d44R94

Specifically, two additional SQL transactions were added to `\Magento\Checkout\Model\GuestPaymentInformationManagement::savePaymentInformationAndPlaceOrder()`, apparently to address "Concurrent checkouts can lead to negative stock" in some way. (It _was not_ added for the registered user version of the code, in `\Magento\Checkout\Model\PaymentInformationManagement::savePaymentInformationAndPlaceOrder()`, which is why only guest checkout is affected. I cannot explain why that might be.)

The effect of the change is that any code (preference, plugin, observer, or otherwise) that triggers a rollback during the `GuestPaymentInformationManagement->savePaymentInformationAndPlaceOrder()` transactions will result in a _"Rolled back transaction has not been completed correctly."_ exception.

The root cause is that Magento's database layer does not appear to be capable of handling rollbacks in the context of nested transactions.

A proper solution would be to either:
1. Remove the added transactions so this situation doesn't arise during order placement, or
2. Fix the database layer to handle rollbacks in this situation successfully.

Versions affected:
2.1.13+
2.2.4+
2.3.0+

**Comment by pixiemediaweb** [comment]
For those having this fun with a newly migrated database - check whether the database has a named definer which hasnt been created in your environment.  Causes this exact same symptom

**Comment by EMBARKwebdesign** [comment]
Hey - I am having a similar issue when creating a new user in the admin are...

```
1 exception(s):
Exception #0 (Exception): Rolled back transaction has not been completed correctly.

Exception #0 (Exception): Rolled back transaction has not been completed correctly.
<pre>#1 Magento\Framework\Model\ResourceModel\AbstractResource->commit() called at [vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php:422]
#2 Magento\Framework\Model\ResourceModel\Db\AbstractDb->save() called at [vendor/magento/framework/Model/AbstractModel.php:655]
#3 Magento\Framework\Model\AbstractModel->save() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#4 Magento\User\Model\User\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#5 Magento\User\Model\User\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#6 Magento\User\Model\User\Interceptor->___callPlugins() called at [generated/code/Magento/User/Model/User/Interceptor.php:832]
#7 Magento\User\Model\User\Interceptor->save() called at [vendor/magento/module-user/Controller/Adminhtml/User/Save.php:96]
#8 Magento\User\Controller\Adminhtml\User\Save->execute() called at [generated/code/Magento/User/Controller/Adminhtml/User/Save/Interceptor.php:24]
#9 Magento\User\Controller\Adminhtml\User\Save\Interceptor->execute() called at [vendor/magento/framework/App/Action/Action.php:108]
#10 Magento\Framework\App\Action\Action->dispatch() called at [vendor/magento/module-backend/App/AbstractAction.php:231]
#11 Magento\Backend\App\AbstractAction->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#12 Magento\User\Controller\Adminhtml\User\Save\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#13 Magento\User\Controller\Adminhtml\User\Save\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-backend/App/Action/Plugin/Authentication.php:143]
#14 Magento\Backend\App\Action\Plugin\Authentication->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#15 Magento\User\Controller\Adminhtml\User\Save\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#16 Magento\User\Controller\Adminhtml\User\Save\Interceptor->___callPlugins() called at [generated/code/Magento/User/Controller/Adminhtml/User/Save/Interceptor.php:39]
#17 Magento\User\Controller\Adminhtml\User\Save\Interceptor->dispatch() called at [vendor/magento/framework/App/FrontController.php:159]
#18 Magento\Framework\App\FrontController->processRequest() called at [vendor/magento/framework/App/FrontController.php:98]
#19 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#20 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#21 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#22 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:26]
#23 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:137]
#24 Magento\Framework\App\Http->launch() called at [generated/code/Magento/Framework/App/Http/Interceptor.php:24]
#25 Magento\Framework\App\Http\Interceptor->launch() called at [vendor/magento/framework/App/Bootstrap.php:261]
#26 Magento\Framework\App\Bootstrap->run() called at [index.php:39]
</pre>
```

Any solutions here please?

**Comment by johnorourke** [comment]
I'm seeing this issue in 2.3.4 when a credit card payment fails, as @rhoerr mentioned above:

`Model/GuestPaymentInformationManagement.php` in module-checkout wraps the placeOrder call with transactions, but the non-guest version doesn't.  Part of the issue is that module-quote/Model/QuoteManagement.php's rollbackAddresses will catch any exception and turn it into a `\Exception` - this doesn't inherit from `\Magento\Framework\Exception\LocalizedException` and so `module-checkout/Model/GuestPaymentInformationManagement.php` turns it into the generic "An error occurred on the server. Please try to place the order again.", hiding the actual message.

Nested transactions aren't supported in mysql AFAIK.

Here's my understanding of where it happens, from reading the 2.3.4 code:
* `\Magento\Checkout\Model\GuestPaymentInformationManagement` begins 2 transactions on different connections
* `\Magento\Quote\Model\QuoteManagement::placeOrder` is called
* `\Magento\Quote\Model\QuoteManagement::submit` is called
* `\Magento\Quote\Model\QuoteManagement::submitQuote` is called
* the `sales_model_service_quote_submit_before` event is dispatched... so far so good
* `\Magento\Sales\Api\OrderManagementInterface::place` is called - say the payment method's `capture` method is called, and it throws a `PaymentException` here because the card details are wrong
* `\Magento\Quote\Model\QuoteManagement::submitQuote` catches it, and calls `rollbackAddresses` (from #21697)
* `\Magento\Quote\Model\QuoteManagement::rollbackAddresses` deletes each address...
* `\Magento\Customer\Model\ResourceModel\Address::delete` calls parent::delete
* `\Magento\Eav\Model\Entity\VersionControl\AbstractEntity::delete` is inherited from...
* `\Magento\Eav\Model\Entity\AbstractEntity::delete` which calls `$this->transactionManager->start($this->getConnection())`
* `\Magento\Eav\Model\Entity\AbstractEntity::getConnection` calls getConnection on the `\Magento\Framework\App\ResourceConnection` it got from the object construction `Context`

So we have two issues:
* `\Magento\Eav\Model\Entity\AbstractEntity` isn't using the Transaction wrapper Magento now provides - but that wouldn't solve it
* having a DB transaction wrapped around a huge deep set of code like order placement will easily cause bugs like this - we should be able to trust the lower-level code (eg. stock changes) to handle errors with transactions, and throw a clean exception with no pending transactions

Checked latest code, and it's fixed in 2.4.1 - see "MC-29426: Orders in Payment System but not in Magento" ( https://github.com/magento/magento2/commit/43e41fac06019d12e955d521cd9034b5954cb0d9 ) which fixed #25862

**Comment by huykon** [comment]
I am facing with this issue when I use custom payment. I don't know how to resolve this, can anyone know to help me?

**Comment by engcom-Hotel** [comment]
Hi @adammprost 

We have tried to reproduce this issue in Magento 2.4-develop, but it seems that this issue has already been resolved. Please try to reproduce this issue in 2.4-develop branch. Please reply on this ticket if you are able to reproduce this issue in this branch.

Thanks

**Comment by engcom-Hotel** [comment]
Hi @adammprost,

It seems that your issue has been resolved. We are unable to get your response since 14 days, hence we are closing this issue for now.

If you still facing the issue, you can reopen the ticket or create a new ticket.

Thanks

**Comment by heqixi** [comment]
I encounter this issue on magento2.4.2,  my reproduce steps are:
1.   create a customer account in my landing page;
2.   log in admin page, create a user (not a customer) account **with the same email address which is used in step 1**,  and then save the user,  the following excepion message was shown

i do not know how to fix, help me please !

----------------------------------------------
1 exception(s):
Exception #0 (Exception): Rolled back transaction has not been completed correctly.

Exception #0 (Exception): Rolled back transaction has not been completed correctly.
<pre>#1 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#2 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#3 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/DB/Adapter/Pdo/Mysql/Interceptor.php:32]
#4 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor->commit() called at [vendor/magento/framework/Model/ResourceModel/AbstractResource.php:93]
#5 Magento\Framework\Model\ResourceModel\AbstractResource->commit() called at [vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php:426]
#6 Magento\Framework\Model\ResourceModel\Db\AbstractDb->save() called at [vendor/magento/framework/Model/AbstractModel.php:655]
#7 Magento\Framework\Model\AbstractModel->save() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#8 Magento\User\Model\User\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#9 Magento\User\Model\User\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#10 Magento\User\Model\User\Interceptor->___callPlugins() called at [generated/code/Magento/User/Model/User/Interceptor.php:581]
#11 Magento\User\Model\User\Interceptor->save() called at [vendor/magento/module-user/Controller/Adminhtml/User/Save.php:96]
#12 Magento\User\Controller\Adminhtml\User\Save->execute() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#13 Magento\User\Controller\Adminhtml\User\Save\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#14 Magento\User\Controller\Adminhtml\User\Save\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php:51]
#15 Magento\Framework\App\Action\Plugin\ActionFlagNoDispatchPlugin->aroundExecute() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#16 Magento\User\Controller\Adminhtml\User\Save\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#17 Magento\User\Controller\Adminhtml\User\Save\Interceptor->___callPlugins() called at [generated/code/Magento/User/Controller/Adminhtml/User/Save/Interceptor.php:23]
#18 Magento\User\Controller\Adminhtml\User\Save\Interceptor->execute() called at [vendor/magento/framework/App/Action/Action.php:111]
#19 Magento\Framework\App\Action\Action->dispatch() called at [vendor/magento/module-backend/App/AbstractAction.php:151]
#20 Magento\Backend\App\AbstractAction->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#21 Magento\User\Controller\Adminhtml\User\Save\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#22 Magento\User\Controller\Adminhtml\User\Save\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-backend/App/Action/Plugin/Authentication.php:143]
#23 Magento\Backend\App\Action\Plugin\Authentication->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#24 Magento\User\Controller\Adminhtml\User\Save\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#25 Magento\User\Controller\Adminhtml\User\Save\Interceptor->___callPlugins() called at [generated/code/Magento/User/Controller/Adminhtml/User/Save/Interceptor.php:32]
#26 Magento\User\Controller\Adminhtml\User\Save\Interceptor->dispatch() called at [vendor/magento/framework/App/FrontController.php:186]
#27 Magento\Framework\App\FrontController->processRequest() called at [vendor/magento/framework/App/FrontController.php:118]
#28 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#29 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#30 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#31 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:23]
#32 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:116]
#33 Magento\Framework\App\Http->launch() called at [generated/code/Magento/Framework/App/Http/Interceptor.php:23]
#34 Magento\Framework\App\Http\Interceptor->launch() called at [vendor/magento/framework/App/Bootstrap.php:263]
#35 Magento\Framework\App\Bootstrap->run() called at [pub/index.php:29]
</pre>

**Comment by savan13** [comment]
I'm facing errors when editing the products.
Can please help me out.
![image](https://user-images.githubusercontent.com/38579373/149919552-8b34d1df-39be-41ec-ac89-f1d7cc99091b.png)

**Comment by savan13** [comment]
> I also had this issue after migrating from Magento 1.9.3 to Magento 2.3. Not sure how to fix it. Is there an update or quick fix to this one? Thank you!!

How will fix that issue? Can you please explain? @josenacasabogiii

**Comment by josenacasabogiii** [comment]
Hi @savan13 

I fixed it by checking the sales order table with some duplicate entries on some increment order id or something like that. You have to check your Magento error logs to point out the specific error because this error is triggered in many different ways when it comes to magento migration. It was luck fixing it on my because I just read some similar issues in in stack Magento exchange and it was long time ago and I can't remember what I specifically did. I'm not really a good Magento developer because I only worked on this platform a few times. 

Just let me know what you found on the errors maybe I can help a bit. Good luck!

**Comment by savan13** [comment]
> Hi @savan13
> 
> I fixed it by checking the sales order table with some duplicate entries on some increment order id or something like that. You have to check your Magento error logs to point out the specific error because this error is triggered in many different ways when it comes to magento migration. It was luck fixing it on my because I just read some similar issues in in stack Magento exchange and it was long time ago and I can't remember what I specifically did. I'm not really a good Magento developer because I only worked on this platform a few times.
> 
> Just let me know what you found on the errors maybe I can help a bit. Good luck!



> Hi @savan13
> 
> I fixed it by checking the sales order table with some duplicate entries on some increment order id or something like that. You have to check your Magento error logs to point out the specific error because this error is triggered in many different ways when it comes to Magento migration. It was luck fixing it on me because I just read some similar issues in stack Magento exchange and it was a long time ago and I can't remember what I specifically did. I'm not really a good Magento developer because I only worked on this platform a few times.
> 
> Just let me know what you found on the errors maybe I can help a bit. Good luck!

@josenacasabogiii 

In error log shows this error: `[2022-01-20 09:43:03] main.CRITICAL: Rolled back transaction has not been completed correctly. {"exception":"[object] (Exception(code: 0): Rolled back transaction has not been completed correctly. at /var/www/html/vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:310)"} []`

Can please look into this?

**Comment by savan13** [comment]
@josenacasabogiii  Hope you are doing well !!

I have found that solution. 
The solution is some 3rd part module database table not defined properly `AUTO INCREMENT & PRIMARY KEY`.
I have checked to disable one-by-one module & check them & got the error.
Now it's working fine.
